feat: add `pgvector` extension and project scaffolding

NiccoloFei opened the pull request:
This commit introduces the `pgvector` extension image and sets up the initial project structure. It also includes generic build, test, and update pipelines to support the future addition of other PostgreSQL extensions.

Closes #2

NiccoloFei committed:
feat: add pgvector

Signed-off-by: Niccolò Fei <niccolo.fei@enterprisedb.com>

ibotty commented on .github/workflows/bake.yml:
```diff
@@ -0,0 +1,56 @@
+        # Remember to add new folders in the operator-changed filter if needed
+        with:
+          base: ${{ (github.event_name == 'schedule') && 'main' || '' }}
+          filters: |
```
Can't we generate this from a glob of a checkout in a previous step?  If so, maybe we could filter on directories that have a `metadata.json`.

NiccoloFei committed:
docs: first draft

Signed-off-by: Niccolò Fei <niccolo.fei@enterprisedb.com>

ibotty commented on test/chainsaw-test.yaml:
```diff
@@ -0,0 +1,28 @@
+  - name: Create a Cluster with the extension
+    try:
+    - apply:
+        file: cluster.yaml
```
If we need to have an extension which has to be in shared_preload_libraries, should this conditionally use either `cluster.yaml` or `cluster-with-shared-preload-library`?

gbartolini replied:
> That would also be easier if it could be e.g. set in the image (as OCI annotation) and read by the operator.

Why would we need that if we are actually the ones producing this information directly at the source? Extending the image catalogue to cover these cases for the extensions will solve the problem in a static and declarative way, controlled at the source.

Moreover, what you are asking requires CNPG to access a registry, and I am not sure we want that (that's not our job, IMHO).

NiccoloFei replied:
Regarding the customization of each extension's GUCS, I've done a first implementation in https://github.com/cloudnative-pg/postgres-extensions-containers/pull/3/commits/764a19d88ca42b0fd8350ff377b6cc9eaed45f8f, and added a way to run more specific tests in case you define them in a `test/` dir in the extension's directory. Ideally I'd also like to automate generating the chainsaw `values.yaml` file locally, to facilitate running the tests locally, but IMHO that's an improvement to be done in a further issue..

NiccoloFei commented on .github/workflows/bake.yml:
```diff
@@ -0,0 +1,56 @@
+        id: get-matrix
+        run: |
+          raw='${{ steps.filter.outputs.changes }}'
+          echo "{\"name\": $raw}" > matrix.json
```
yeah, quoting https://github.com/dorny/paths-filter/tree/de90cc6fb38fc0963ad72b210f1f284cd68cea36/?tab=readme-ov-file#outputs:
```
changes - JSON array with names of all filters matching any of the changed files.
```

So if multiple extensions get updated together, they'll all be part of the matrix passed to the `Bake` job, which will spawn a new Job for each extension.
Although the idea is to limit the release cycle of each extension to their own component owners, so that extensions don't interfere as much with each other, and you don't have to bother fixing unrelated failures if you are just changing your extension.

ibotty replied:
But won't `matrix.name` below (line 56) be an array then?  We'll see when another extension gets built.

NiccoloFei replied:
That output looks like
```
echo 'matrix={"name": ["pgvector"]}'
```

So the job strategy will be like
```yaml
strategy:
  matrix:
    name: [pgvector]
```

and with more extensions
```yaml
strategy:
  matrix:
    name: [pgvector, postgis]
```

so `matrix.name` will just be the name of the current extension. Am I missing something here?

NiccoloFei committed:
ci: fix template expansion

Signed-off-by: Niccolò Fei <niccolo.fei@enterprisedb.com>

NiccoloFei committed:
ci: QOL changes

Signed-off-by: Niccolò Fei <niccolo.fei@enterprisedb.com>

NiccoloFei committed:
test: allow templating extension's GUCS

Signed-off-by: Niccolò Fei <niccolo.fei@enterprisedb.com>

NiccoloFei committed:
chore: add pgvector license

Signed-off-by: Niccolò Fei <niccolo.fei@enterprisedb.com>

fcanovai committed:
fix: set base image to scratch

The org.opencontainers.image.base.name should not reference images used
to generate the contents of the image. The base image we have in the
FROM statement is scratch.

Signed-off-by: Francesco Canovai <francesco.canovai@enterprisedb.com>

fcanovai committed:
fix: reorder image tags to prioritize extension version

Switch extension version and PostgreSQL major in the tag.
Compatibilities (pg major and distro) are conventionally at the end of a
name.

Signed-off-by: Francesco Canovai <francesco.canovai@enterprisedb.com>

NiccoloFei committed:
ci: remove outdated comment

Signed-off-by: Niccolò Fei <niccolo.fei@enterprisedb.com>

mnencia committed:
chore: review

Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

gbartolini committed:
docs: review

Signed-off-by: Gabriele Bartolini <gabriele.bartolini@enterprisedb.com>
